### PR TITLE
Refactor empty function

### LINKED ISSUE #114: Refactor empty function
Reimplement the empty function as part of the refactored design of the v3 implementation.

This should also close #97.

### PR DESCRIPTION
**Pull request by johngrimes**
Closes #114.

### CODE REVIEW

File: fhir-server/src/main/java/au/csiro/pathling/fhirpath/function/EmptyFunction.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+  @Nonnull
+  @Override
+  public FhirPath invoke(@Nonnull final NamedFunctionInput input) {
+    checkNoArguments(NAME, input);
+    final NonLiteralPath inputPath = input.getInput();
+    final String expression = expressionFromInput(input, "empty");
+
+    // "Empty" means that the group contains only null values.
+    final Function<Column, Column> empty = col -> when(max(col).isNull(), true).otherwise(false);
```
Comment (piotrszul) [inline fhir-server/src/main/java/au/csiro/pathling/fhirpath/function/EmptyFunction.java:39]: I am not certain if all columns types we use are comparable, even if they are maybe a more efficient expression would be:
`COUNT(col) == 0`

File: fhir-server/src/main/java/au/csiro/pathling/fhirpath/function/EmptyFunction.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import java.util.function.Function;
+import javax.annotation.Nonnull;
+import org.apache.spark.sql.Column;
+import org.hl7.fhir.r4.model.Enumerations.FHIRDefinedType;
+
+/**
+ * This function returns true if the input collection is empty.
+ *
+ * @author John Grimes
+ * @see <a href="https://pathling.csiro.au/docs/fhirpath/functions.html#empty">first</a>
```
Comment (piotrszul) [inline fhir-server/src/main/java/au/csiro/pathling/fhirpath/function/EmptyFunction.java:25]: empty NOT first (in the link name)

File: fhir-server/src/main/java/au/csiro/pathling/QueryHelpers.java [hunk 3]
```diff
@@ -132,18 +136,44 @@ public static DatasetWithIdAndValue convertRawResource(@Nonnull final Dataset<Ro
...
-    final Column leftId = checkPresent(left.getIdColumn());
-    final Column rightId = checkPresent(right.getIdColumn());
 
-    final Column joinCondition = leftId.equalTo(rightId);
-    final Dataset<Row> target = selectJoinTarget(right.getDataset(), left.getDataset());
-    return left.getDataset().join(target, joinCondition, joinType.getSparkName());
+    final Column leftId;
+    final Column rightId;
+    leftId = checkPresent(left.getIdColumn());
+    rightId = checkPresent(right.getIdColumn());
```
Comment (piotrszul) [inline fhir-server/src/main/java/au/csiro/pathling/QueryHelpers.java:152]: Initialise at declaration.

**Review by piotrszul — COMMENTED** [review]
Looks good to me! 
Only some minor tweaks.

File: fhir-server/src/main/java/au/csiro/pathling/fhirpath/function/EmptyFunction.java (hunk 1 above)
Comment (johngrimes) [reply fhir-server/src/main/java/au/csiro/pathling/fhirpath/function/EmptyFunction.java]: See eaf03dc1c9d930006fe2050ce346dc83062a142e.

File: fhir-server/src/main/java/au/csiro/pathling/fhirpath/function/EmptyFunction.java (hunk 2 above)
Comment (johngrimes) [reply fhir-server/src/main/java/au/csiro/pathling/fhirpath/function/EmptyFunction.java]: See 124fd90795650e34b41eb302bf3f0e94d743da4f.

File: fhir-server/src/main/java/au/csiro/pathling/QueryHelpers.java (hunk 3 above)
Comment (johngrimes) [reply fhir-server/src/main/java/au/csiro/pathling/QueryHelpers.java]: See d0dcdba3a30ccb8b1a2698dd99c8f82c5c81b5da.

**Pull request merged**